[UNIT] Final review of Unit 2.1 on smolagents - release 25/2/25 (#195)
huggingface/agents-course: units/en/unit2/introduction.mdx

> # Agentic Frameworks

Reviewer: ```suggestion
# Introduction to Agentic Frameworks
```

> 

Reviewer: ```suggestion

<img src="https://huggingface.co/datasets/agents-course/course-images/resolve/main/en/unit2/general-thumbnail.jpg" alt="General Thumbnail"/>
```

Reviewer: (This is the placeholder for the general thumbnail of unit 2)

Author: I think we should add a real banner not a placeholder so it's solved.

> After introducing the core concepts behind agents, we'll explore different agentic frameworks that can be used to build powerful agentic applications. Let's dive in! 🕵

Reviewer: ```suggestion
Welcome to this second unit, where **we'll explore different agentic frameworks** that can be used to build powerful agentic applications. 

We will study:

- In Unit 2.1: [Smolagents](https://huggingface.co/docs/smolagents/en/index)
- In Unit 2.2: [LlamaIndex](https://www.llamaindex.ai/)
- In Unit 2.3: [LangGraph](https://www.langchain.com/langgraph)

Let's dive in! 🕵
```

## When to Use an Agentic Framework

> An agentic framework is not always needed when building an application around LLMs. They provide flexibility in the workflow to efficiently solve a specific task, but they're not always necessary. Sometimes, predefined workflows are sufficient to fulfill user requests, and there is no real need for an agentic framework. If the approach to build an agent is simple, like a chain of prompts, using plain code may be enough. The advantage is that the developer will have full control and understanding of their system without abstractions.

Reviewer: ```suggestion
An agentic framework is **not always needed when building an application around LLMs**. They provide flexibility in the workflow to efficiently solve a specific task, but they're not always necessary. 

Sometimes, **predefined workflows are sufficient** to fulfill user requests, and there is no real need for an agentic framework. If the approach to build an agent is simple, like a chain of prompts, using plain code may be enough. The advantage is that the developer will have **full control and understanding of their system without abstractions**.
```

Reviewer: The changes are making here is adding a little bit of bold and space because the content is very good but it's big paragraphs.

<img width="723" alt="Screenshot 2025-02-21 at 06 43 33" src="https://github.com/user-attachments/assets/d2b5f85f-f02d-4f80-bdb0-8ab798d9f435" />

However, when the workflow becomes more complex, such as letting an LLM call functions or using multiple agents, these abstractions start to become helpful.

Considering these ideas, we can already identify the need for some features:

> * An LLM engine that powers the system.

Reviewer: ```suggestion
* An *LLM engine* that powers the system.
```

> * A list of tools the agent can access.

Reviewer: ```suggestion
* A *list of tools* the agent can access.
```

> * A parser for extracting tool calls from the LLM output.

Reviewer: ```suggestion
* A *parser* for extracting tool calls from the LLM output.
```

* A system prompt synced with the parser.

> * A memory system.

Reviewer: ```suggestion
* A *memory system*.
```

> * Error logging and retry mechanisms to control LLM mistakes.

Reviewer: ```suggestion
* *Error logging and retry mechanisms* to control LLM mistakes.
```
